orangecontrib/text/widgets/owbagofwords.py

```diff
@@ -1,126 +1,127 @@
+                self,
+                'autocommit',
+                'Commit',
+                box=False
+        )
+
+    def set_data(self, data):
         self.corpus = data
+        self.commit()
+
+    def commit(self):
         self.apply()
 
     def apply(self):
-        # TODO Move the logic to the scripting module.
-        new_table = None
-        if self.corpus:
-            if self.preprocessor:
-                if self.use_tfidf:
-                    cv = TfidfVectorizer(lowercase=self.preprocessor.lowercase,
-                                         stop_words=self.preprocessor.stop_words,
-                                         preprocessor=self.preprocessor.transformation,
-                                         norm=self.normalization)
-                else:
-                    cv = CountVectorizer(lowercase=self.preprocessor.lowercase,
-                                         stop_words=self.preprocessor.stop_words,
-                                         preprocessor=self.preprocessor.transformation)
-
-                pp_info_tag = "Includes punctuation: {}\nLowercase: {}\n" \
-                              "Transformation: {}\n" \
-                              "Stop words removed: {}\n" \
-                              "TF-IDF performed: {}\n" \
-                              "Normalization: {}"\
-                    .format(self.preprocessor.incl_punct, self.preprocessor.lowercase,
-                            str(self.preprocessor.transformation),
-                            self.preprocessor.stop_words, self.use_tfidf,
-                            self._normalization_options[self.normalization_type])
-                self.pp_info_label.setText(pp_info_tag)
-            else:
-                cv = CountVectorizer(lowercase=True,
-                                     binary=True)
-
-            documents = self.corpus.documents
-            feats = cv.fit(documents)  # Features.
-            freqs = feats.transform(documents).toarray()  # Frequencies.
-
-            # Generate the domain attributes.
-            attr = [ContinuousVariable(f) for f in feats.get_feature_names()]
-
-            # Construct a new domain.
-            domain = Domain(attr, self.corpus.domain.class_vars, metas=self.corpus.domain.metas)
-
-            # Create the table.
-            new_table = Table.from_numpy(domain, freqs, Y=self.corpus._Y, metas=self.corpus.metas)
-        self.send("Data", new_table)
-
-    def _select_normalization(self, n):
-        if n > 0:
-            self.normalization = self._normalization_options[n].split(" ")[0].lower()
+        self.error(0, '')
+        if self.corpus is not None:
+            # BoW uses 4 steps ATM.
+            # Update this when a better solution is available.
+            with self.progressBar(4) as progress_bar:
+                self.progress_bar = progress_bar
+                bag_of_words = BagOfWords(
+                        progress_callback=self.advance_progress,
```

Comment:
`progress_callback=progress_bar.advance` seems to work just fine. There is no need for storing the progress bar into `self.progress_bar` and method `advance_progress` could also be removed.

This also fixes `UserWarning: progressBarSet() called without a preceding progressBarInit()` that is now raised when you click on the `commit` button for the second time.